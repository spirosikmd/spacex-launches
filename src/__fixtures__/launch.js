export function createLaunchResponse(data) {
  return {
    flight_number: 1,
    mission_name: 'Test',
    launch_date_utc: '2019-01-01T18:16:00.000Z',
    launch_success: true,
    upcoming: false,
    is_tentative: false,
    details: 'Details',
    mission_id: ['id1'],
    links: {
      mission_patch: 'missionPatch',
    },
    ...data,
  };
}

export function createLaunch(data) {
  return {
    flightNumber: 1,
    missionName: 'Test',
    utcDate: new Date('2019-01-01'),
    isSuccessful: true,
    isFailed: false,
    isUpcoming: false,
    isInProgress: false,
    isTentative: false,
    details: 'Details',
    missionIds: ['id1'],
    missionPatch: 'missionPatch',
    ...data,
  };
}
