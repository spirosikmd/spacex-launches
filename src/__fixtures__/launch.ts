import { LaunchResponseData, LaunchData } from '../api';

export function createLaunchResponse(
  data: LaunchResponseData
): LaunchResponseData {
  return {
    flight_number: 1,
    mission_name: 'Test',
    launch_date_utc: '2019-01-01T18:16:00.000Z',
    launch_success: true,
    upcoming: false,
    is_tentative: false,
    details: 'Details',
    mission_id: ['id1'],
    links: {
      mission_patch: 'missionPatch',
    },
    ...data,
  };
}

export function createLaunch(data: LaunchData): LaunchData {
  return {
    flightNumber: 1,
    missionName: 'Test',
    utcDate: new Date('2019-01-01'),
    isSuccessful: true,
    isFailed: false,
    isUpcoming: false,
    isInProgress: false,
    isTentative: false,
    details: 'Details',
    missionIds: ['id1'],
    missionPatch: 'missionPatch',
    ...data,
  };
}
