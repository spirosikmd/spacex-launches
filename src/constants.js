export const ASC = 'asc';
export const DESC = 'desc';
export const FLIGHT_NUMBER_FIELD = 'flight_number';
export const UTC_DATE_FIELD = 'launch_date_utc';
export const FIELDS = `${FLIGHT_NUMBER_FIELD},${UTC_DATE_FIELD},launch_success,upcoming,is_tentative,details,mission_name,mission_id`;
