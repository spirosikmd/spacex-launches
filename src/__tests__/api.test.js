import {
  getLaunch,
  getLaunches,
  mapSortField,
  LAUNCHES_BASE,
  FILTER,
} from '../api';
import { ASC, FLIGHT_NUMBER_FIELD, UTC_DATE_FIELD } from '../constants';
import { createLaunchResponse } from '../__fixtures__/launch';

describe('getLaunch', () => {
  beforeEach(() => {
    fetch.resetMocks();
  });

  it('returns successful launch data', async () => {
    fetch.mockResponseOnce(JSON.stringify(createLaunchResponse()));

    const launch = await getLaunch({ flightNumber: '1' });

    expect(launch).toMatchSnapshot();
    expect(fetch.mock.calls.length).toEqual(1);
    expect(fetch).toBeCalledWith(`${LAUNCHES_BASE}/1`, {
      headers: { 'Content-Type': 'application/json' },
    });
  });

  it('returns failed launch data', async () => {
    fetch.mockResponseOnce(
      JSON.stringify(createLaunchResponse({ launch_success: false }))
    );

    const launch = await getLaunch({ flightNumber: '1' });

    expect(launch.isFailed).toBe(true);
  });

  it('returns upcoming launch data', async () => {
    fetch.mockResponseOnce(
      JSON.stringify(
        createLaunchResponse({ launch_success: null, upcoming: true })
      )
    );

    const launch = await getLaunch({ flightNumber: '1' });

    expect(launch.isUpcoming).toBe(true);
  });

  it('returns in progress launch data', async () => {
    fetch.mockResponseOnce(
      JSON.stringify(
        createLaunchResponse({ launch_success: null, upcoming: false })
      )
    );

    const launch = await getLaunch({ flightNumber: '1' });

    expect(launch.isInProgress).toBe(true);
  });

  it('returns launch data with empty mission ids', async () => {
    fetch.mockResponseOnce(
      JSON.stringify(createLaunchResponse({ mission_id: null }))
    );

    const launch = await getLaunch({ flightNumber: '1' });

    expect(launch.missionIds.length).toBe(0);
  });

  it('returns launch data without mission patch', async () => {
    fetch.mockResponseOnce(
      JSON.stringify(createLaunchResponse({ links: { mission_patch: null } }))
    );

    const launch = await getLaunch({ flightNumber: '1' });

    expect(launch.missionPatch).toBe('');
  });

  it('throws error', async () => {
    fetch.mockResponseOnce(JSON.stringify({ error: 'Not Found' }), {
      status: 404,
    });

    try {
      await getLaunch({ flightNumber: '1' });
    } catch (error) {
      expect(error.message).toBe('Not Found');
    }
  });
});

describe('getLaunches', () => {
  beforeEach(() => {
    fetch.resetMocks();

    fetch.mockResponseOnce(
      JSON.stringify([
        createLaunchResponse(),
        createLaunchResponse({ flight_number: 2 }),
      ])
    );
  });

  it('returns sanitized launches data', async () => {
    const response = await getLaunches({
      sortOrder: ASC,
      sortField: FLIGHT_NUMBER_FIELD,
    });

    expect(response).toMatchSnapshot();

    expect(fetch.mock.calls.length).toEqual(1);
    expect(fetch).toBeCalledWith(
      `${LAUNCHES_BASE}?filter=${FILTER}&sort=flight_number&order=asc`,
      {
        headers: { 'Content-Type': 'application/json' },
      }
    );
  });

  it('throws error', async () => {
    fetch.mockResponseOnce(JSON.stringify({ error: 'Not Found' }), {
      status: 404,
    });

    try {
      await getLaunches({
        sortOrder: ASC,
        sortField: FLIGHT_NUMBER_FIELD,
      });
    } catch (error) {
      expect(error.message).toBe('Not Found');
    }
  });
});

describe('mapSortField', () => {
  it('returns "flight_number" for flightNumber', () => {
    expect(mapSortField(FLIGHT_NUMBER_FIELD)).toBe('flight_number');
  });

  it('returns "launch_date_utc" for utcDate', () => {
    expect(mapSortField(UTC_DATE_FIELD)).toBe('launch_date_utc');
  });
});
