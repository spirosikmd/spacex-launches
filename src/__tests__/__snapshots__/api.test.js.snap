// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getLaunch returns successful launch data 1`] = `
Object {
  "details": "Details",
  "flightNumber": 1,
  "isFailed": false,
  "isInProgress": false,
  "isSuccessful": true,
  "isTentative": false,
  "isUpcoming": false,
  "missionIds": Array [
    "id1",
  ],
  "missionName": "Test",
  "missionPatch": "missionPatch",
  "utcDate": Object {
    "getDate": [MockFunction],
    "getFullYear": [MockFunction],
    "getMonth": [MockFunction],
    "toUTCString": [MockFunction],
  },
}
`;

exports[`getLaunches returns sanitized launches data 1`] = `
Array [
  Object {
    "details": "Details",
    "flightNumber": 1,
    "isFailed": false,
    "isInProgress": false,
    "isSuccessful": true,
    "isTentative": false,
    "isUpcoming": false,
    "missionIds": Array [
      "id1",
    ],
    "missionName": "Test",
    "missionPatch": "missionPatch",
    "utcDate": Object {
      "getDate": [MockFunction],
      "getFullYear": [MockFunction],
      "getMonth": [MockFunction],
      "toUTCString": [MockFunction],
    },
  },
  Object {
    "details": "Details",
    "flightNumber": 2,
    "isFailed": false,
    "isInProgress": false,
    "isSuccessful": true,
    "isTentative": false,
    "isUpcoming": false,
    "missionIds": Array [
      "id1",
    ],
    "missionName": "Test",
    "missionPatch": "missionPatch",
    "utcDate": Object {
      "getDate": [MockFunction],
      "getFullYear": [MockFunction],
      "getMonth": [MockFunction],
      "toUTCString": [MockFunction],
    },
  },
]
`;
